### ISSUE TITLE
Versioning feedback & documentation for different path to achieve “no app binding”

### ISSUE THREAD

**Issue by durera** [redhat-developer/service-binding-operator#1001]
https://github.com/redhat-developer/service-binding-operator/issues/897 states:

> Will need to provide documentation about a different path to achieve the “same use case”

Where can we find this documentation?  This use case is the key use case in our product and we are trying to work out how we continue to operate under version 0.9.0 schema, is there an example somewhere of how to achieve the same thing in 0.9 that you can point us to?

Also, really disappointed that this change was introduced without a rev to the CRD API version, it is going to make it very difficult to diagnose and debug issues in any environment we are using that there is now two versions of the v1alpha API.  We appreciate this is pre 1.0 release and there will be breaking changes in the API, but we would expect that the standard ways of handling breaking API changes would still be applied in the lead up to v1, e.g. v1alpha1, v1alpha2, and so on.

The lack of proper versioning is making our lives as early adopters of this project far more difficult than it needs to be and I would ask that you please rev the API schema in future breaking changes so that we can have clean and clear upgrade paths as the project evolves instead of being stuck in a position of trying to work out “well, is it the 0.8 v1alpha or the 0.9 v1alpha”.

It would also be really helpful if could make available channels that aligned to api version changes, so that we can easily control the upgrade as we move through breaking changes, such that we could have automatically picked up 0.7 -> 0.8 upgrade on an alpha1 channel, but the 0.9 release would have not been made available on that channel, so that anyone with auto-upgrade turned on isn’t instantly broken, and could choose to move to alpha2 only once their own application code has been updated to be compatible with the API changes.

I really think it’s in the best interests of the project to consider the early adopters’ experience consuming this a little more closely, as there have been many avoidable pain points like this where there are well established operating patterns designed precisely to avoid these being a problem in the first place.  We really like what you are putting together here and see it as a key technology for us now and in the future, but the lax versioning is really starting to hurt the teams here that are depending on this project.

**Comment by pedjak** [comment]
> Also, really disappointed that this change was introduced without a rev to the CRD API version, it is going to make it very difficult to diagnose and debug issues in any environment we are using that there is now two versions of the v1alpha API. We appreciate this is pre 1.0 release and there will be breaking changes in the API, but we would expect that the standard ways of handling breaking API changes would still be applied in the lead up to v1, e.g. v1alpha1, v1alpha2, and so on.

Changing API version would impact all consumers, even those not affected by the dropped functionality. Given that we are still in pre 1.0 mode, supporting multiple API versions is something that we do not have a bandwidth for. However, I can agree that we could release 0.9.0 on a different channel, to avoid automatic updates. Apologize for that.

> Where can we find this documentation? This use case is the key use case in our product and we are trying to work out how we continue to operate under version 0.9.0 schema, is there an example somewhere of how to achieve the same thing in 0.9 that you can point us to?

Can you please share some details about the use case and why SBO should not inject bindings into the application? Dropping the functionality was announced on [March 4 SBO community call](https://docs.google.com/document/d/1HwhAKqpM6l4Ur3h3IApDFzbH2Y_xvj_n1x1pEdwRuSY/edit) as a part of aligning to [the spec](https://github.com/k8s-service-bindings/spec). SBO is more than a secret generator. If you really need just a secret containing all binding data, perhaps creating a controller that creates/manages [Provisioned Service](https://github.com/k8s-service-bindings/spec#provisioned-service) resources is better tailored for your usecase?

**Comment by durera** [comment]
Sorry for the long delay in responding, various things have been preventing me getting back to this.

The use case is where the binding is to a complex application with many microservices. As such we don't want SBO directly modifying the deployments, we have an operator that manages the various microservices that are impacted by a change in the binding.

- The binding is owned by the operator.
- When processing an operand, the operator creates a binding to one or more services on behalf of that operand
- When one of the services we have bound the operand to updates, SBO will update the binding resource
- The operator (because it is watching the binding resource) will now reconcile the operand
- The operator can now apply the change across the workloads that represent the operand in a controlled manner, and if required perform additional steps to aid in the change of configuration (e.g. put an app into maintenance mode/perform an ordered restart of workloads to apply change etc).  
- If SBO directly modifies the workloads it will be taking control away from the operator, we don't want that.

Seems a huge backwards step to us.  In a simple single workload application this may make sense, but in complex systems where there are operators involved tying everything together dropping this capability makes it impossible to continue to use SBO without giving up control in the operator, unless I'm missing something SBO will just immediately modify all deployments referenced in the binding - taking away the ability for the operator to do it's job.

> SBO is more than a secret generator. If you really need just a secret containing all binding data, perhaps creating a controller that creates/manages Provisioned Service resources is better tailored for your usecase?

Yeah, we could create a new controller that does all the same things, sans the "bind to a workload" part.  We felt it would be a bad idea to do so however, and that it made more sense at the time to adopt this technology because SBO was exactly what we needed that controller to do.  

If you're saying that this use case is no longer supported I guess that's what we will ultimately have to do that, we can fork from 0.8 or start from scratch, but I got the impression from this that the usecase is still supported, there's just some different path to achieve it?

> Will need to provide documentation about a different path to achieve the “same use case”

I think we were using SBO as far more than secret generator and it's value lies not purely in making the modification to the workload; SBO was the broker between services in other namespaces that we wanted to bind a system to, without giving that system permissions outside it's own namespace.  This felt like, and still feels like, a primary use case for a service binding operator to me.

- service exists in namespace1
- application exists in namespace2
- SBO runs in cluster scope
- application just needs to know the name of the thing it wants to bind to and what namespace it lives in.  It doesn't need any cluster permissions, it simply needs RBAC to create a bindingrequest in it's own namespace and SBO will serve up all the necessary information for the application to bind to the service.
- Whether the operator applies the result of the binding, or whether SBO does it directly is immaterial imo.

The fact that we want the operator to ultimately control the application of the binding to workloads doesn't lessen the role SBO is performing imo. Looking at the link you provided I don't think that really helps as it's within-namespace, so isn't really solving the same problem; unless I'm misreading it.

Back in Feb we shared details of the MAS & SBO usecase on one of the community calls .. I've attached some of the slides that covered one of the examples of how SBO is used.

![image](https://user-images.githubusercontent.com/4400618/132690296-a7f6c58b-3f8e-4b89-b3ce-d14d011ea68b.png)
![image](https://user-images.githubusercontent.com/4400618/132690322-0d0e5f44-297e-4580-a851-6a38e8323ce6.png)
![image](https://user-images.githubusercontent.com/4400618/132690354-1b71aec7-0545-4b45-b037-6c2a571be37c.png)
![image](https://user-images.githubusercontent.com/4400618/132690384-1dcce5b1-c9ae-4296-b1d2-3a02249de9f7.png)

## tldr
**tldr; I guess ultimately we just need to know whether there is a way to achieve the same thing with SBO, because if not we are going to need to crack on with producing a replacement that does support this usecase.**

We have had half a dozen support tickets coming in weekly since 0.9 was released, we've publishing a tech note on the subject: [IBM Maximo Application Suite 8.4.x and 8.5.x require Service-Binding operator 0.8](https://www.ibm.com/support/pages/node/6477000)

I see two paths forward for our Feb 2022 release:
- If this use case won't be supported by SBO, abandon use of SBO and start development of a replacement technology that is able to stop at the "binding secret available to use" stage.
- If this use case will be supported by SBO, In the short term we can look at creating a simple dummy workload which we don't really care about that SBO can be placed in charge of for the purpose of fulfilling the binding until such time that a release is available that supports this again.

**Comment by pedjak** [comment]
@durera thanks for a very detailed feedback, we are going to discuss it during [our the next community call](https://github.com/redhat-developer/service-binding-operator/#community-discussion-contribution-and-support), feel free to join it.

**Comment by pedjak** [comment]
> Seems a huge backwards step to us. In a simple single workload application this may make sense, but in complex systems where there are operators involved tying everything together dropping this capability makes it impossible to continue to use SBO without giving up control in the operator, unless I'm missing something SBO will just immediately modify all deployments referenced in the binding - taking away the ability for the operator to do it's job.

In this case, it seems to me natural that you model this complex system with a custom resource (if I see it correctly from the diagram, that would a resource representing `Suite`). That resource will be managed by your operator and SBO can inject binding secret reference into that resource on a given place. Your operator can detect that and then start rolling out updates to appropriate components.

I am going to try to sketch how this could look like. Imagine `Suite` resource like this:

```yaml
apiVersion: example.com/v1
kind: Suite
spec:
  application1:
     bindingSecret: ""
```

Following binding can be created:
```yaml
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
    name: binding1
spec:
    application:
        name: suite
        group: example.com
        version: v1
        kind: Suite
        bindingPath:
            secretPath: spec.application1.bindingSecret
    services:
      - group: example.com
        version: v1
        kind: Service
        namespace: service-ns
        name: service1            
```

and as the result, `Suite` instance will contain the reference to the binding secret. That could be repeated for all applications, if needed.

Would this work for you? We have also [an acceptance test demonstrating it](https://github.com/redhat-developer/service-binding-operator/blob/master/test/acceptance/features/injectSecretAtCustomSchemaPath.feature#L79).

We dropped "no app binding" usecase as a part of aligning with [the service binding specification](https://github.com/k8s-service-bindings/spec). Would you like to open an issue there, so that we discuss it further?

**Comment by johnpoth** [comment]
Hi @durera ! Coincidentally, we were also using SBO in the same manner (for some of the same reasons) and were blocked when support for this feature was removed.

As a workaround, not sure if it's applicable to you, we now SBO as a [library](https://github.com/apache/camel-k/blob/main/pkg/trait/service_binding.go#L65) inside our operator. It's fairly easy to [remove](https://github.com/apache/camel-k/blob/main/pkg/trait/service_binding.go#L166) the check for " no app binding" and stop at secret generation. Thanks !
